src/idl/tests/annotation.c

```diff
@@ -242,44 +244,98 @@ CU_Test(idl_annotation, default_literal)
   }
 }
 
-CU_Test(idl_annotation, key)
-{
+typedef struct key_test {
+  const char *str;
   idl_retcode_t ret;
+  bool val[8];
+  bool annotated[8];
+} key_test_t;
+
+static void test_key(key_test_t test)
+{
   idl_pstate_t *pstate = NULL;
-  idl_struct_t *s;
-  idl_member_t *m;
-  const char str[] = "struct s {\n"
-                     "  @key char a;\n"
-                     "  @key(TRUE) char b;\n"
-                     "  @key(FALSE) char c;\n"
-                     "  char d;\n"
-                     "};";
 
-  ret = parse_string(IDL_FLAG_ANNOTATIONS, str, &pstate);
-  CU_ASSERT_EQUAL_FATAL(ret, IDL_RETCODE_OK);
+  idl_retcode_t ret = parse_string(IDL_FLAG_ANNOTATIONS, test.str, &pstate);
+
+  CU_ASSERT_EQUAL(ret, test.ret);
+
+  if (ret)
+    return;
+
   CU_ASSERT_PTR_NOT_NULL_FATAL(pstate);
   assert(pstate);
-  s = (idl_struct_t *)pstate->root;
-  CU_ASSERT_FATAL(idl_is_struct(s));
-  assert(s);
-  m = (idl_member_t *)s->members;
-  CU_ASSERT_FATAL(idl_is_member(m));
-  CU_ASSERT_PTR_NOT_NULL(m->key.annotation);
-  CU_ASSERT(m->key.value == true);
-  assert(m);
-  m = idl_next(m);
-  CU_ASSERT_FATAL(idl_is_member(m));
-  CU_ASSERT_PTR_NOT_NULL(m->key.annotation);
-  CU_ASSERT(m->key.value == true);
-  m = idl_next(m);
-  CU_ASSERT_FATAL(idl_is_member(m));
-  CU_ASSERT_PTR_NOT_NULL(m->key.annotation);
-  CU_ASSERT(m->key.value == false);
-  m = idl_next(m);
-  CU_ASSERT_FATAL(idl_is_member(m));
-  CU_ASSERT_PTR_NULL(m->key.annotation);
-  CU_ASSERT(m->key.value == false);
+
+  if (idl_is_struct(pstate->root)) {
+    idl_struct_t *s = (idl_struct_t *)pstate->root;
+    assert(s);
+    size_t i = 0;
+    idl_member_t *m = NULL;
+    IDL_FOREACH(m, s->members) {
+      if (test.annotated[i]) {
+        CU_ASSERT_PTR_NOT_NULL(m->key.annotation);
+      } else {
+        CU_ASSERT_PTR_NULL(m->key.annotation);
+      }
+      CU_ASSERT(m->key.value == test.val[i]);
+
+      i++;
+    }
+  } else if (idl_is_union(pstate->root)) {
+    idl_union_t *u = (idl_union_t *)pstate->root;
+    assert(u);
+
+    if (test.annotated[0]) {
+      CU_ASSERT_PTR_NOT_NULL(u->switch_type_spec->key.annotation);
+    } else {
+      CU_ASSERT_PTR_NULL(u->switch_type_spec->key.annotation);
+    }
+    CU_ASSERT(u->switch_type_spec->key.value == test.val[0]);
+  } else {
+    CU_ASSERT(false);
+  }
+
   idl_delete_pstate(pstate);
+
+}
+
+CU_Test(idl_annotation, key)
+{
+  key_test_t tests[] = {
+    {"struct s {\n"
+     "  @key char a;\n"
+     "  @key(TRUE) char b;\n"
+     "  @key(FALSE) char c;\n"
+     "  @must_understand @key char d;\n"
+     "  @optional(FALSE) @key char e;\n"
```

Comment:
To avoid overlap in the `@optional`, `@key` and `@must-understand` tests with respect to annotation order, you could consider using the primary annotation for the test (`@key` in this case) as the first annotation and include all combinations of the other to as 2nd and 3rd. This ensures that all paths in the `annotate_...` functions are tested.

Reply:
I went with the following approach:
The first annotation to be parsed, and therefore to be set is the one that occurs first in the text.
So to check for the code paths checking for conflicting annotations in the `annotate_key` function, which in my eyes is the purpose of the test, I first set the conflicting annotation(s) `@optional` and `@must_understand` and then `@key` second.